### PR TITLE
feat: option to disable record reordering

### PR DESCRIPTION
**Pull request by kinolaev**
I'd like to have an option to prioritize the latest event during deduplication, because it seems like debezium events are already ordered. For example, for a sequence of operations
```sql
insert into t1(id, text) values(1, 'hello');
delete from t1 where id = 1;
insert into t1(id, text) values(1, 'hallo');
```
I'd like to get a row `1,'hallo'` in an iceberg table. But because deletes have higher priority than inserts, I get `1,'hello'`.

I'm not sure that events are always ordered and that is why I propose to add an option instead of changing the default behavior.

### CODE REVIEW

File: debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/IcebergTableOperator.java [hunk 1]
```diff
@@ -60,13 +61,33 @@ protected List<EventConverter> deduplicateBatch(List<EventConverter> events) {
...
+                Long ts = e.cdcSourceTsValue();
+                Operation prevOp = prev.cdcOpValue();
+                Operation op = e.cdcOpValue();
+                if (Long.compare(prevTs, ts) > 0 ||
+                    (prevOp == Operation.INSERT && op == Operation.INSERT) ||
+                    (prevOp == Operation.UPDATE && op == Operation.INSERT) ||
+                    (prevOp == Operation.DELETE && op == Operation.UPDATE) ||
+                    (prevOp == Operation.DELETE && op == Operation.DELETE)
+                ) {
+                  LOGGER.warn("Out of order events for key {} detected: {} {} -> {} {}", e.key(), prevTs, prevOp, ts, op);
```
Comment (ismailsimsek) [inline debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/IcebergTableOperator.java]: could we create a test to trigger this situation?  IcebergTableOperatorTest.testDeduplicateBatch has a way to create manual events and de-duplicate them.

File: debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/IcebergTableOperator.java [hunk 2]
```diff
@@ -57,53 +57,15 @@ protected List<EventConverter> deduplicateBatch(List<EventConverter> events) {
...
-   * delete. for this case we need to only pick last change event for the row.
-   * <p>
-   * Its used when `upsert` feature enabled (when the consumer operating non append mode) which means it should not add
-   * duplicate records to target table.
-   *
-   * @param lhs
-   * @param rhs
-   * @return
-   */
-  private int compareByTsThenOp(EventConverter lhs, EventConverter rhs) {
```
[the remark this answers is not in the document]
Comment (kinolaev) [reply debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/IcebergTableOperator.java]: Done 6470820ca0dcf1c886b2581a278e4fbafae3148b

File: debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java [hunk 3]
```diff
@@ -50,7 +50,8 @@ InternalRecordWrapper wrapper() {
   public void write(Record row) throws IOException {
     RowDataDeltaWriter writer = route(row);
     Operation rowOperation = ((RecordWrapper) row).op();
-    if (rowOperation == Operation.INSERT) {
+    boolean isNewKey = ((RecordWrapper) row).isNewKey();
+    if (isNewKey && !keepDeletes && rowOperation != Operation.DELETE) {
       // new row
       writer.write(row);
     } else if (rowOperation == Operation.DELETE && !keepDeletes) {
```
Comment (ismailsimsek) [inline debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java]:
```suggestion
        // Hard DELETE: The record is deleted and history is disabled - remove the key without writing a tombstone.
    } else if (rowOperation == Operation.DELETE && !keepDeletes) {
```

File: debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java
Comment (ismailsimsek) [inline debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java]:
```suggestion
        // Upsert/Update: Key already exists (Update) or Soft Delete - delete the old version first to ensure uniqueness/deduplication, then write the new state.
    } else {
```

File: debezium-server-iceberg-sink/src/test/java/io/debezium/server/iceberg/IcebergChangeConsumerUpsertTest.java [hunk 4]
```diff
@@ -96,19 +96,6 @@ public void testSimpleUpsert() throws Exception {
     Assertions.assertEquals(ds.where("id = 5 AND __op= 'r' ").count(), 1);
     Assertions.assertEquals(ds.where("id = 6 AND __op= 'u' AND first_name= 'Updatedname-6-V1'").count(), 1);
```
Comment (ismailsimsek) [inline debezium-server-iceberg-sink/src/test/java/io/debezium/server/iceberg/IcebergChangeConsumerUpsertTest.java]: is there any additional or new edge cases to add ?

File: debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java [hunk 5]
```diff
@@ -50,7 +50,8 @@ InternalRecordWrapper wrapper() {
   public void write(Record row) throws IOException {
     RowDataDeltaWriter writer = route(row);
     Operation rowOperation = ((RecordWrapper) row).op();
-    if (rowOperation == Operation.INSERT) {
+    boolean isNewKey = ((RecordWrapper) row).isNewKey();
+    if (isNewKey && !keepDeletes && rowOperation != Operation.DELETE) {
```
[the remark this answers is not in the document]
Comment (kinolaev) [reply debezium-server-iceberg-sink/src/main/java/io/debezium/server/iceberg/tableoperator/BaseDeltaTaskWriter.java]: Added in 533c1ceee1bf629f1cc7ba2d3b4b90ae3d311829

File: docs/iceberg.md [hunk 6]
```diff
@@ -63,13 +63,13 @@ mode.
 
 #### Upsert Mode Data Deduplication
 
-Upsert mode enables data deduplication, prioritizing records based on their `__source_ts_ns` timestamp and operation
-type (`__op`). The `debezium.sink.iceberg.upsert-dedup-column` property can be used to specify a different column for
-deduplication (currently limited to Long type).
+Upsert mode enables data deduplication. By default only the latest record in a batch is retained for each primary key.
```
[the remark this answers is not in the document]
Comment (ismailsimsek) [reply docs/iceberg.md]: great, thank you for checking it.
